src/tool/xmeta/cpp/ast_to_st_listener.cpp

```diff
@@ -0,0 +1,206 @@
+             e_member.resolve_decimal_val(new_enum->get_type());
+             new_enum->add_member(std::move(e_member));
+         }
+         else
+         {
+             enum_member e_member{ enum_member_id, decl_line, "" };
+             auto const& val = new_enum->get_members().back().get_value();
+             if (!val.is_resolved() && val.get_ref_name() == e_member.get_id())
+             {
+                 m_st.write_enum_circular_dependency(
+                     e_member.get_decl_line(),
+                     e_member.get_id(),
+                     new_enum->get_id());
+                 return false;
+             }
+             e_member.set_value(val);
+             ec = e_member.increment(new_enum->get_type());
+             new_enum->add_member(std::move(e_member));
+         }
+     }
+
+     if (ec == std::errc::result_out_of_range)
+     {
+         m_st.write_enum_const_expr_range_error(
+             decl_line,
+             expr->getText(),
+             enum_member_id);
+         return false;
+     }
+     else
+     {
+         assert(ec == std::errc());
+     }
+     return true;
+}
+
+bool ast_to_st_listener::resolve_enum_val(enum_member& e_member, std::shared_ptr<enum_model> const& new_enum, std::set<std::string_view>& dependents)
+{
+    auto const& val = e_member.get_value();
+    if (val.is_resolved())
+    {
+        return true;
+    }
+    else
+    {
+        if (dependents.find(e_member.get_id()) != dependents.end())
+        {
+            m_st.write_enum_circular_dependency(
+                e_member.get_decl_line(),
+                e_member.get_id(),
+                new_enum->get_id());
+                return false;
+        }
+        dependents.emplace(e_member.get_id());
+        auto const& ref_name = val.get_ref_name();
+        if (!new_enum->member_exists(ref_name))
+        {
+            m_st.write_enum_member_expr_ref_error(e_member.get_decl_line(), ref_name, new_enum->get_id());
+            return false;
+        }
+        auto ref_member = new_enum->get_member(ref_name);
+        if (!resolve_enum_val(ref_member, new_enum, dependents))
+        {
+            return false;
+        }
+        e_member.set_value(ref_member.get_value());
+    }
+    return true;
+}
+
+void ast_to_st_listener::exitEnum_declaration(XlangParser::Enum_declarationContext *ctx)
+{
+    auto id = ctx->IDENTIFIER();
+    auto decl_line = id->getSymbol()->getLine();
+    enum_semantics type = enum_semantics::Int32;
+
+    if (ctx->enum_base())
+    {
+        assert(ctx->enum_base()->enum_integral_type() != nullptr);
+        type = enum_semantics_map[ctx->enum_base()->enum_integral_type()->getText()];
+    }
+
+    std::shared_ptr<enum_model> new_enum = std::make_shared<enum_model>(id->getText(), decl_line, m_st.get_cur_assembly(), type);
+
+    for (auto field : ctx->enum_body()->enum_member_declaration())
+    {
+        if (!extract_enum_member(field, new_enum))
+        {
+            return;
```

Comment:
if this fails, we should stop the listener in its tracks and try to output a exception or something to indicate it failed